src/tensor/diag.jl

```diff
@@ -31,13 +31,13 @@ Base.eltype(::Type{<:Diag{T}}) where {T} = eltype(T)
 # Deal with uniform Diag conversion
 Base.convert(::Type{<:Diag{T}},D::Diag) where {T} = Diag{T}(data(D))
 
-Base.similar(D::Diag{T}) where {T} = Diag{T}(similar(data(D)))
+Base.similar(D::Diag{T}, ::Type{T}) where {T} = Diag{T}(similar(data(D)))
 
 # TODO: write in terms of ::Int, not inds
-Base.similar(D::Diag{T},inds) where {T} = Diag{T}(similar(data(D),minimum(dims(inds))))
-Base.similar(D::Type{<:NonuniformDiag{T}},inds) where {T} = Diag{T}(similar(T,length(inds)==0 ? 1 : minimum(dims(inds))))
+Base.similar(D::Diag{T},inds::IndexSet) where {T} = Diag{T}(similar(data(D),minimum(dims(inds))))
```

Comment:
Was there a reason for adding `IndexSet` here? The goal is actually to have all of the code in the `tensor/` directory independent of other `ITensor` types (like ITensor, Index, IndexSet, etc.). Eventually we want to turn it into a stand-alone library, so `inds` here is meant to be generic (either an IndexSet with no QNs, a `NTuple{N,Int}`, or anything else with `dims` defined on it).

Reply:
Without it, the code tries to call this function for `::Diag{Float64}, Float64` as args and `Float64` has no method `dims`.

Reply:
That sounds like a `UniformDiag` case, again there it sounds like we should have a dispatch branch in `similar` for `UniformDiag` and `NonuniformDiag`.

Reply:
Although with the new method, it's not necessary anymore it seems, so I can remove that type requirement too.